flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sources/FieldComputer.scala

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.sources
+
+import org.apache.flink.api.common.typeinfo.TypeInformation
+import org.apache.flink.table.api.ValidationException
+import org.apache.flink.table.expressions.{Expression, ResolvedFieldReference}
+
+/**
+  * The [[FieldComputer]] interface returns an expression to compute the field of the table schema
+  * of a [[TableSource]] from one or more fields of the [[TableSource]]'s return type.
+  *
+  * @tparam T The result type of the provided expression.
+  */
+abstract class FieldComputer[T] {
```

Comment:
I think we could add a `getArguments: Array[_]` function to allow providing extra arguments besides the existing fields, e.g., for situations in which event-time is represented as `String`, the argument should be the time pattern.

Reply:
constant attributes can be passed via the constructor (see `ExistingField`).
For example you can define your example as 

```
class ParseStringField(field: String, format: String) extends TimestampExtractor
```

The format string is directly embedded into the generated expression.